server/src/kafka.rs

```diff
@@ -0,0 +1,231 @@
+    }
+    if let Ok(val) = env::var("KAFKA_GROUP") {
+        conf.set("group.id", &val);
+    }
+
+    if let Ok(val) = env::var("KAFKA_SECURITY_PROTOCOL") {
+        let mapped: SslProtocol = val.parse()?;
+        conf.set("security.protocol", &mapped.to_string());
+    }
+    let consumer: StreamConsumer = conf.create()?;
+
+    if let Ok(vals_raw) = env::var("KAFKA_PARTITIONS") {
+        let vals = vals_raw
+            .split(',')
+            .map(i32::from_str)
+            .collect::<Result<Vec<i32>, ParseIntError>>()
+            .map_err(|raw| KafkaError::ParseIntError("KAFKA_PARTITIONS", raw))?;
+
+        let mut parts = TopicPartitionList::new();
+        for val in vals {
+            parts.add_partition(&topic, val);
+        }
+        consumer.seek_partitions(parts, Timeout::Never)?;
+    }
+    Ok(consumer)
+}
+
+fn resolve_schema(stream_name: &str) -> Result<HashMap<String, Arc<Field>>, KafkaError> {
+    let hash_map = STREAM_INFO.read().unwrap();
+    let raw = hash_map
+        .get(stream_name)
+        .ok_or_else(|| KafkaError::StreamNotFound(stream_name.to_owned()))?;
+    Ok(raw.schema.clone())
+}
+async fn ingest_message<'a>(stream_name: &str, msg: BorrowedMessage<'a>) -> Result<(), KafkaError> {
+    log::debug!("{}: Message: {:?}", stream_name, msg);
+    if let Some(payload) = msg.payload() {
+        let schema = resolve_schema(stream_name)?;
+        let event = format::json::Event {
+            data: serde_json::from_slice(payload)?,
+            tags: String::default(),
+            metadata: String::default(),
+        };
+        log::debug!("Generated event: {:?}", event.data);
+        let (rb, is_first) = event.into_recordbatch(schema, None, None).unwrap();
+
+        event::Event {
+            rb,
+            stream_name: stream_name.to_string(),
+            origin_format: "json",
+            origin_size: payload.len() as u64,
+            is_first_event: is_first,
+            parsed_timestamp: Utc::now().naive_utc(),
+            time_partition: None,
+            custom_partition_values: HashMap::new(),
+            stream_type: StreamType::UserDefined,
+        }
+        .process()
+        .await?;
+    } else {
+        log::debug!("{} No payload for stream", stream_name);
+    }
+    Ok(())
+}
+
+pub async fn setup_integration() -> Result<JoinHandle<()>, KafkaError> {
+    let my_res = if let Ok(stream_name) = env::var("KAFKA_TOPIC") {
+        log::info!("Setup kafka integration for {stream_name}");
+        create_stream_if_not_exists(&stream_name, &StreamType::UserDefined.to_string()).await?;
+
+        let res = task::spawn(async move {
+            let consumer = setup_consumer().unwrap();
+            let mut stream = consumer.stream();
+            loop {
+                while let Some(curr) = stream.next().await {
+                    let msg = curr.unwrap();
+                    ingest_message(&stream_name, msg).await.unwrap();
+                }
+            }
```

Comment:
`while let` inside a `loop`? Also, why the unwrap is not after `stream.next()` and instead on the returned value when [ref](https://github.com/fede1024/rust-rdkafka/blob/master/src/consumer/stream_consumer.rs#L317) says to do otherwise?